packages/api/src/middleware/maintenance.js

```diff
@@ -0,0 +1,97 @@
+import { HTTPError } from '../errors.js'
+
+/**
+ * @typedef {'rw' | 'r-' | '--'} Mode
+ * @typedef {import('../utils/router.js').Handler} Handler
+ */
+
+/**
+ * Read and write.
+ */
+export const READ_WRITE = 'rw'
+
+/**
+ * Read only mode.
+ */
+export const READ_ONLY = 'r-'
+
+/**
+ * No reading or writing.
+ */
+export const NO_READ_OR_WRITE = '--'
+
+/** @type {readonly Mode[]} */
+export const modes = Object.freeze([NO_READ_OR_WRITE, READ_ONLY, READ_WRITE])
+
+/**
+ * The default maintenance mode (normal operation).
+ */
+export const DEFAULT_MODE = READ_WRITE
+
+/** @type {() => Mode} */
+let getMaintenanceMode = () => DEFAULT_MODE
+
+/**
+ * Sets the function that returns the current maintenance mode. This allows us
+ * to pass a function that accesses a global variable. Cloudflare secrets are
+ * exposed as global variables so this is the way to dynamically access them.
+ *
+ * The return value from the getter function dictates which request handlers
+ * (that have been wrapped with `withMode`) are enabled. The returned value is
+ * one of:
+ *
+ * -- = no reading or writing
+ * r- = read only mode
+ * rw = read and write (normal operation)
+ *
+ * @param {() => Mode} getter
+ */
+export function setMaintenanceModeGetter(getter) {
+  getMaintenanceMode = getter
+}
```

Comment:
this can't mutate DEFAULT_MODE directly ?

Reply:
I don't understand the question...